File: Programme_C/chapitre-6/exo-4.c
/*
Écrire une programme qui charge en mémoire un tableau d’entiers tel que généré au a). Le
fichier d’entiers ne contient pas le nombre d’éléments. Le programme doit fonctionner pour un
nombre quelconque de données entières dans le fichier.
*/
#include <stdio.h>
#include <stdlib.h>
#include <fcntl.h>
#include <unistd.h>

int *lecture_binaire(const char *fichier, int *taille) 
{
    int fd = open(fichier, O_RDONLY) ;
    
    if (fd == -1) 
    {
        perror("Erreur ouverture fichier") ;
        return NULL;
    }

    int capacite = 10 ;          
    int *tab = malloc(capacite * sizeof(int)) ;

    if (!tab) 
    {
        perror("Erreur allocation memoire") ;
        close(fd) ;
        return NULL ;
    }

    int n = 0 ;  
    int buffer ;
    ssize_t bytes ;

    while ((bytes = read(fd, &buffer, sizeof(int))) == sizeof(int)) 
    {
        if (n >= capacite) 
        {
            capacite *= 2 ; 

            int *tmp = realloc(tab, capacite * sizeof(int)) ;
            if (!tmp) 
            {
                perror("Erreur realloc") ;
                free(tab) ;
                close(fd) ;
                return NULL ;
            }

            tab = tmp ;
        }

        tab[n++] = buffer ;
    }

    if (bytes == -1) 
    {
        perror("Erreur lecture fichier");
        free(tab);
        close(fd);
        return NULL;
    }

    close(fd);
    *taille = n ; 
    return tab;
}

int main() 
{
     char nom_fichier[] = "tableau.bin" ;
     int taille ;
     int *tab = lecture_binaire(nom_fichier, &taille) ;

     if (!tab) 
     {
        perror("ERREUR DE RECUPERATION") ;
        return 1 ;
     }
 
     int i = 0 ;
     
     printf("Tableau charge (%d elements) :\n", taille) ;
     for (i = 0 ; i < taille ; i++) 
     {
        printf("%d ", tab[i]) ;
     }

     printf("\n") ;

     free(tab) ;
     return 0;
}

File: Programme_Python/fichier/exo-4.py
import struct
import os

def lecture_binaire(fichier):
    tab = []

    if not os.path.isfile(fichier):
        print(f"Erreur : le fichier '{fichier}' n'existe pas.")
        return None

    try:
        with open(fichier, "rb") as f:
            while True:
                bytes_read = f.read(4)  # taille d'un int (4 octets)
                if not bytes_read:
                    break
                if len(bytes_read) != 4:
                    print("Erreur : taille du bloc lue incorrecte")
                    return None
                # 'i' = int 4 octets, format standard C
                val = struct.unpack('i', bytes_read)[0]
                tab.append(val)
    except OSError as e:
        print(f"Erreur lecture fichier : {e}")
        return None

    return tab


if __name__ == "__main__":
    nom_fichier = "tableau.bin"
    tab = lecture_binaire(nom_fichier)

    if tab is None:
        print("ERREUR DE RECUPERATION")
    else:
        print(f"Tableau chargé ({len(tab)} éléments) :")
        print(" ".join(str(x) for x in tab))